CLI (pmcp status / pmcp update) can't reach own gateway: ClientManager always uses SSE transport regardless of config.type

## Summary

`pmcp status` and `pmcp update <server>` hang / fail against a running PMCP gateway v1.9.2 on `http://127.0.0.1:3344/mcp`, even though the gateway is healthy (other MCP clients — Claude Code, `mcp__pmcp__gateway_health` — work fine against the same endpoint).

The root cause is a **client-side transport dispatch bug**: `ClientManager._connect_server` unconditionally calls `_connect_sse` for any `RemoteMcpServerConfig`, regardless of whether the config declares `type="http"`, `"streamable-http"`, or `"sse"`. Since the gateway's `/mcp` route is streamable-HTTP (via `StreamableHTTPSessionManager`) and not the legacy SSE dual-endpoint transport, `mcp.client.sse.sse_client` hits a protocol mismatch — either a hard 400 (on 1.8.x) or an indefinite hang waiting for a never-arriving SSE `endpoint` event (on 1.9.2, due to the rmcp-compat keep-alive stream introduced for #51).

This also triggers a **server-side ASGI protocol exception** (`Expected ASGI message 'http.response.body', but got 'http.response.start'`) when the failed client's request eventually hits the 60s timeout in `handle_mcp`.

## Environment

- PMCP 1.9.2 (PyPI, installed via `pip install --user`)
- Gateway running as `pmcp --transport http --host 127.0.0.1 --port 3344`
- systemd `pmcp.service` (user), gateway healthy: `/health` returns `{"ok":true,"version":"1.9.2","transport":"http"}`
- Other MCP clients (Claude Code with `.mcp.json` type=http, Gemini) connect successfully via streamable-HTTP

## Repro

With a healthy gateway running:

**On 1.9.2 — silent hang:**
```
$ pmcp update context7
(hangs indefinitely, no output; requires ^C)
```
The SSE client connects, gets 200 OK + the rmcp-compat keep-alive stream (added in 1.9.2 as a workaround for #51), and waits forever for an `endpoint` SSE event that never arrives — the server is streamable-HTTP, not SSE, so there is no endpoint event to send.

**On 1.8.x — explicit failure:**
```
$ pmcp update context7
[INFO] HTTP Request: GET http://127.0.0.1:3344/mcp "HTTP/1.1 400 Bad Request"
[WARNING] Connection to pmcp-gateway failed (attempt 1/3), retrying in 1.0s
[WARNING] Connection to pmcp-gateway failed (attempt 2/3), retrying in 2.0s
[ERROR] Failed to connect to pmcp-gateway: unhandled errors in a TaskGroup
Fatal error: Unknown tool: pmcp-gateway::gateway.update_server
```
Before the rmcp keep-alive workaround existed, the session manager rejected session-less GETs with 400 outright, surfacing the transport mismatch loudly.

**In both cases** `pmcp status` silently falls back to static config scan (exception swallowed in `_query_running_gateway_status`'s broad `except Exception: return None`). The gateway logs `RuntimeError: Expected ASGI message 'http.response.body', but got 'http.response.start'` after the 60s `handle_mcp` timeout.

**Interesting interaction with #51:** the rmcp-compat keep-alive SSE added to help Codex quietly converted PMCP's own CLI failure from a loud 3-retries-then-error into a silent indefinite hang. Fixing the root transport-dispatch bug below clears both symptoms without touching the rmcp workaround.

## Root Cause

### 1. CLI client is transport-blind (primary)

`src/pmcp/client/manager.py:383-389`:

```python
async def _connect_server(self, config: ResolvedServerConfig) -> None:
    """Connect to a single MCP server."""
    if isinstance(config.config, RemoteMcpServerConfig):
        await self._connect_sse(config)   # <-- always SSE, regardless of config.type
        return
    await self._connect_stdio(config)
```

`RemoteMcpServerConfig.type: Literal["remote", "sse", "http", "streamable-http"]` (`types.py:44`) advertises four transport options, but the manager only ever invokes `_connect_sse`, which calls `mcp.client.sse.sse_client(...)`. There is no dispatch to `streamablehttp_client` from `mcp.client.streamable_http`.

Against a streamable-HTTP server, the SSE client's opening GET to `/mcp` either (a) on 1.8.x, receives a 400 because the session manager rejects session-less GETs, or (b) on 1.9.2, receives a 200 + rmcp-compat keep-alive SSE (`: keep-alive\n\n` comments only, no `endpoint` event). In case (b) the SSE client hangs forever waiting for the `endpoint` event it needs to know where to POST.

The `PMCP_STATUS_SSE_URL` env-var name in `cli.py` (used by both `run_status` and `run_update`) is a legacy hint that the CLI predates the streamable-HTTP migration.

### 2. Server ASGI double-start on timeout (secondary)

`src/pmcp/transport/http.py` `handle_mcp`:

```python
try:
    await asyncio.wait_for(
        session_manager.handle_request(request.scope, request.receive, request._send),
        timeout=request_timeout,
    )
except asyncio.TimeoutError:
    return Response("Gateway Timeout", status_code=504)   # <-- BUG
_inc("requests_ok")
return _NullResponse()
```

The `_NullResponse` sentinel exists specifically because `session_manager.handle_request` writes to `send` directly — returning a real `Response` from Starlette's endpoint path would double-send. But on timeout, we **do** return a real `Response(504)`. If the session manager already sent `http.response.start` (e.g. to open an SSE stream) before we hit the timeout, Starlette then tries to send another `http.response.start` → uvicorn raises `RuntimeError: Expected ASGI message 'http.response.body', but got 'http.response.start'.`

### 3. CLI swallows connection failures (tertiary)

`run_update` in `src/pmcp/cli.py:663`:

```python
await client_manager.connect_all([gateway_config])
try:
    ...
    result = await _call("gateway.update_server", {"server_name": server_name})
```

`connect_all` returns a list of error messages rather than raising. When the gateway connection fails all 3 retries, `connect_all` returns errors (logged at ERROR level) but the update code proceeds to call the tool anyway, producing `Unknown tool: pmcp-gateway::gateway.update_server`. The user-facing error is misleading and misattributes the failure.

`run_status` has the same issue but explicitly catches and returns None (line 786), so the symptom is a silent fallback rather than a confusing error — still wrong, but different user experience.

### 4. Test gap

`tests/test_cli.py` mocks `_query_running_gateway_status` to return `None` and never exercises a real gateway handshake. `tests/test_client_manager.py` covers stdio and SSE but no streamable-HTTP client connection. This is why the 1.9.2 release passed CI despite the transport mismatch.

## Proposed Fix

### Primary — client transport dispatch

In `ClientManager._connect_server`, dispatch based on `config.config.type`:

```python
async def _connect_server(self, config: ResolvedServerConfig) -> None:
    if isinstance(config.config, RemoteMcpServerConfig):
        transport_type = config.config.type
        if transport_type in ("http", "streamable-http"):
            await self._connect_streamable_http(config)
        else:  # "sse" or legacy "remote"
            await self._connect_sse(config)
        return
    await self._connect_stdio(config)
```

Add `_connect_streamable_http` that mirrors `_connect_sse` but uses `mcp.client.streamable_http.streamablehttp_client`. Default type (`"remote"`) could either preserve SSE for back-compat or sniff the server via a HEAD/GET to `/health` and auto-select; I'd vote for an explicit deprecation warning on `"remote"` and a migration path to `"streamable-http"`.

Update `cli.py` `run_status` and `run_update` to explicitly set `type="streamable-http"` on the `gateway_config` they construct (since the gateway's `/mcp` is always streamable-http). The legacy env var should probably be renamed `PMCP_GATEWAY_URL`, keeping `PMCP_STATUS_SSE_URL` as a deprecated alias.

### Secondary — ASGI double-start

Wrap `request._send` with a tracking shim before passing to `session_manager.handle_request`:

```python
response_started = False
async def tracking_send(message):
    nonlocal response_started
    if message["type"] == "http.response.start":
        response_started = True
    await request._send(message)

try:
    await asyncio.wait_for(
        session_manager.handle_request(request.scope, request.receive, tracking_send),
        timeout=request_timeout,
    )
except asyncio.TimeoutError:
    if response_started:
        # Session manager already wrote response start — can't send another.
        # Close the connection rather than double-starting.
        return _NullResponse()
    return Response("Gateway Timeout", status_code=504)
```

### Tertiary — CLI error handling

`run_update` should inspect the return value of `connect_all` and bail with a clear message on failure:

```python
errors = await client_manager.connect_all([gateway_config])
if errors:
    print(f"Error: cannot reach PMCP gateway at {gateway_url}", file=sys.stderr)
    for err in errors:
        print(f"  {err}", file=sys.stderr)
    print("Hint: check 'systemctl --user status pmcp' or 'curl http://127.0.0.1:3344/health'", file=sys.stderr)
    sys.exit(1)
```

Same treatment in `run_status`'s `_query_running_gateway_status` — log the actual exception at INFO/DEBUG so users can troubleshoot when the live-query fallback silently kicks in.

### Tests

Add an integration test that spins up the real gateway in streamable-HTTP mode (`create_http_app` + `uvicorn`), has the CLI client connect and call `gateway.health`, and asserts the handshake succeeds. This would have caught the regression.

## Gaps / capabilities worth considering (while in the area)

1. **Bootstrap doesn't auto-upgrade PMCP.** `uv tool install pmcp` is a no-op if already installed; fleets drift (we found 1.5.1 / 1.8.0 / 1.8.1 / 1.9.2 across four hosts until manual intervention). Consider `bootstrap.sh` running `uv tool upgrade pmcp || pip install -U --user --break-system-packages pmcp` when `--upgrade` flag or version-check logic triggers.
2. **`pip install --user` and `uv tool install` both target `~/.local/bin/pmcp`** — whichever ran last wins, creating silent drift between the uv tool venv and the active binary. Bootstrap picks uv; manual installs often pick pip. `pmcp doctor` could surface this conflict.
3. **No `pmcp upgrade` subcommand for pmcp-itself.** Could call `uv tool upgrade pmcp` or `pip install -U pmcp` based on how it was installed (detected from `sys.executable` / `~/.local/share/uv/tools/pmcp`). Useful complement to `pmcp update <server>`.
4. **`pmcp doctor` could probe the gateway's transport** (`/health` returns `transport: "http"`) and compare against the configured client type, catching exactly this misconfiguration class.
5. **Related closed issue #51** (Codex streamable-HTTP warnings) likely shares root cause — Codex's rmcp client wasn't handling PMCP's session lifecycle, which was worked around server-side. Our case is the mirror: PMCP's own client not speaking streamable-HTTP at all.

Happy to send a PR for items 1-3 above if that's useful — they're contained and each has a clean test.

--- a/.gitignore
+++ b/.gitignore
@@ -38,6 +38,7 @@ env/
 htmlcov/
 .tox/
 .nox/
+.benchmarks/
 
 # OS
 .DS_Store
@@ -45,6 +46,8 @@ Thumbs.db
 
 # Logs
 *.log
+.pmcp/
+.playwright-mcp/
 
 # Env files
 .env
@@ -57,3 +60,7 @@ Thumbs.db
 # BAML generated client
 src/mcp_gateway/baml_client/
 data.db
+code_index.db
+
+# Local screenshot artifacts
+/*.png

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -4,7 +4,7 @@ build-backend = "hatchling.build"
 
 [project]
 name = "pmcp"
-version = "1.9.2"
+version = "1.9.3"
 description = "PMCP - Progressive MCP: Minimal context bloat with on-demand tool discovery"
 readme = "README.md"
 license = "MIT"

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -36,7 +36,7 @@ dependencies = [
     "pyyaml>=6.0",
     "python-dotenv>=1.0.0",
     "aiohttp>=3.9.0",
-    "baml-py==0.219.0",
+    "baml-py==0.221.0",
 ]
 
 [project.urls]

--- a/src/pmcp/__init__.py
+++ b/src/pmcp/__init__.py
@@ -1,3 +1,3 @@
 """PMCP - A meta-server for minimal Claude Code tool bloat."""
 
-__version__ = "1.9.2"
+__version__ = "1.9.3"

--- a/src/pmcp/cli.py
+++ b/src/pmcp/cli.py
@@ -685,11 +685,11 @@ async def run_update(args: argparse.Namespace) -> None:
         max_tools_per_server=policy_manager.get_max_tools_per_server()
     )
 
-    gateway_url = os.environ.get("PMCP_STATUS_SSE_URL", "http://127.0.0.1:3344/mcp")
+    gateway_url = _get_gateway_url()
     gateway_config = ResolvedServerConfig(
         name="pmcp-gateway",
         source="custom",
-        config=RemoteMcpServerConfig(type="http", url=gateway_url),
+        config=RemoteMcpServerConfig(type="streamable-http", url=gateway_url),
     )
 
     async def _call(tool_name: str, arguments: dict[str, object]) -> dict[str, object]:
@@ -698,8 +698,11 @@ async def run_update(args: argparse.Namespace) -> None:
         )
         return _extract_tool_payload(raw) or {}
 
-    await client_manager.connect_all([gateway_config])
     try:
+        errors = await client_manager.connect_all([gateway_config])
+        if errors:
+            _exit_gateway_unreachable(errors)
+
         targets: list[str] = []
         if args.all:
             health = await _call("gateway.health", {})
@@ -738,6 +741,21 @@ async def run_update(args: argparse.Namespace) -> None:
         await client_manager.disconnect_all()
 
 
+def _get_gateway_url() -> str:
+    """Return the PMCP gateway MCP endpoint URL."""
+    return os.environ.get(
+        "PMCP_GATEWAY_URL",
+        os.environ.get("PMCP_STATUS_SSE_URL", "http://127.0.0.1:3344/mcp"),
+    )
+
+
+def _exit_gateway_unreachable(errors: list[str]) -> None:
+    """Exit with a clear direct-gateway connection failure."""
+    detail = "; ".join(errors)
+    print(f"Error: cannot reach PMCP gateway: {detail}", file=sys.stderr)
+    sys.exit(1)
+
+
 def _extract_tool_payload(result: dict[str, object]) -> dict[str, object] | None:
     """Extract JSON payload from MCP tool response content."""
     content = result.get("content")
@@ -762,24 +780,29 @@ def _extract_tool_payload(result: dict[str, object]) -> dict[str, object] | None
 async def _query_running_gateway_status(
     args: argparse.Namespace, policy_manager: object
 ) -> dict[str, object] | None:
-    """Query live gateway status over SSE, return None on failure."""
+    """Query live gateway status over streamable HTTP, return None on failure."""
     import time
 
     from pmcp.client.manager import ClientManager
     from pmcp.types import RemoteMcpServerConfig, ResolvedServerConfig
 
+    logger = logging.getLogger(__name__)
     probe_manager = ClientManager(
         max_tools_per_server=policy_manager.get_max_tools_per_server()  # type: ignore[attr-defined]
     )
-    probe_url = os.environ.get("PMCP_STATUS_SSE_URL", "http://127.0.0.1:3344/mcp")
+    probe_url = _get_gateway_url()
     probe_config = ResolvedServerConfig(
         name="pmcp-gateway",
         source="custom",
-        config=RemoteMcpServerConfig(type="http", url=probe_url),
+        config=RemoteMcpServerConfig(type="streamable-http", url=probe_url),
     )
 
     try:
-        await probe_manager.connect_all([probe_config])
+        errors = await probe_manager.connect_all([probe_config])
+        if errors:
+            logger.debug("Live gateway status query failed: %s", "; ".join(errors))
+            return None
+
         health_raw = await probe_manager.call_tool(
             "pmcp-gateway::gateway.health", {}, timeout_ms=3000
         )
@@ -821,6 +844,7 @@ async def _query_running_gateway_status(
 
         return snapshot
     except Exception:
+        logger.debug("Live gateway status query failed", exc_info=True)
         return None
     finally:
         await probe_manager.disconnect_all()
@@ -1772,11 +1796,11 @@ async def run_auth_connect(args: argparse.Namespace) -> None:
         max_tools_per_server=policy_manager.get_max_tools_per_server()
     )
 
-    gateway_url = os.environ.get("PMCP_STATUS_SSE_URL", "http://127.0.0.1:3344/mcp")
+    gateway_url = _get_gateway_url()
     gateway_config = ResolvedServerConfig(
         name="pmcp-gateway",
         source="custom",
-        config=RemoteMcpServerConfig(type="http", url=gateway_url),
+        config=RemoteMcpServerConfig(type="streamable-http", url=gateway_url),
     )
 
     async def _call(tool_name: str, arguments: dict[str, object]) -> dict[str, object]:
@@ -1786,8 +1810,11 @@ async def run_auth_connect(args: argparse.Namespace) -> None:
         payload = _extract_tool_payload(raw)
         return payload or {}
 
-    await client_manager.connect_all([gateway_config])
     try:
+        errors = await client_manager.connect_all([gateway_config])
+        if errors:
+            _exit_gateway_unreachable(errors)
+
         server_name = args.server_name
         first = await _call("gateway.provision", {"server_name": server_name})
 

--- a/src/pmcp/client/manager.py
+++ b/src/pmcp/client/manager.py
@@ -18,6 +18,7 @@ from typing import Any
 
 import mcp.types as mcp_types
 from mcp.client.sse import sse_client
+from mcp.client.streamable_http import streamablehttp_client
 from mcp.shared.message import SessionMessage
 
 from pmcp.config.loader import make_tool_id
@@ -191,6 +192,15 @@ def _interpolate_header_value(value: str) -> str:
     if not match:
         return value
     return os.environ.get(match.group(1), "")
+
+
+def _remote_headers(config: RemoteMcpServerConfig) -> dict[str, str] | None:
+    """Return remote transport headers with env-var placeholders expanded."""
+    if not config.headers:
+        return None
+    return {
+        key: _interpolate_header_value(value) for key, value in config.headers.items()
+    }
 
 
 @dataclass
@@ -383,7 +393,10 @@ class ClientManager:
     async def _connect_server(self, config: ResolvedServerConfig) -> None:
         """Connect to a single MCP server."""
         if isinstance(config.config, RemoteMcpServerConfig):
-            await self._connect_sse(config)
+            if config.config.type in ("http", "streamable-http"):
+                await self._connect_streamable_http(config)
+            else:
+                await self._connect_sse(config)
             return
 
         await self._connect_stdio(config)
@@ -573,6 +586,38 @@ class ClientManager:
 
     async def _connect_sse(self, config: ResolvedServerConfig) -> None:
         """Connect to a remote SSE MCP server."""
+        if not isinstance(config.config, RemoteMcpServerConfig):
+            raise ValueError(f"Server {config.name} has unsupported remote config type")
+
+        remote_config = config.config
+        headers = _remote_headers(remote_config)
+        await self._connect_remote_stream(
+            config,
+            sse_client(remote_config.url, headers=headers),
+            transport_name="SSE",
+        )
+
+    async def _connect_streamable_http(self, config: ResolvedServerConfig) -> None:
+        """Connect to a remote streamable-HTTP MCP server."""
+        if not isinstance(config.config, RemoteMcpServerConfig):
+            raise ValueError(f"Server {config.name} has unsupported remote config type")
+
+        remote_config = config.config
+        headers = _remote_headers(remote_config)
+        await self._connect_remote_stream(
+            config,
+            streamablehttp_client(remote_config.url, headers=headers),
+            transport_name="streamable HTTP",
+        )
+
+    async def _connect_remote_stream(
+        self,
+        config: ResolvedServerConfig,
+        transport_context: Any,
+        *,
+        transport_name: str,
+    ) -> None:
+        """Connect to a remote MCP server using a read/write stream transport."""
         name = config.name
 
         status = ServerStatus(
@@ -582,29 +627,17 @@ class ClientManager:
         )
         self._servers[name] = status
 
-        if not isinstance(config.config, RemoteMcpServerConfig):
-            raise ValueError(f"Server {name} has unsupported remote config type")
+        logger.info(f"Connecting to remote MCP server via {transport_name}: {name}")
 
-        remote_config = config.config
-        headers = None
-        if remote_config.headers:
-            headers = {
-                key: _interpolate_header_value(value)
-                for key, value in remote_config.headers.items()
-            }
-
-        logger.info(f"Connecting to remote MCP server: {name}")
-
-        sse_stack = AsyncExitStack()
-        read_stream, write_stream = await sse_stack.enter_async_context(
-            sse_client(remote_config.url, headers=headers)
-        )
+        remote_stack = AsyncExitStack()
+        transport = await remote_stack.enter_async_context(transport_context)
+        read_stream, write_stream = transport[:2]
 
         managed = ManagedClient(
             config=config,
             process=None,
             is_remote=True,
-            sse_exit_stack=sse_stack,
+            sse_exit_stack=remote_stack,
             write_stream=write_stream,
             status=status,
         )
@@ -717,7 +750,7 @@ class ClientManager:
         except Exception as e:
             status.status = ServerStatusEnum.ERROR
             status.last_error = str(e)
-            await sse_stack.aclose()
+            await remote_stack.aclose()
             raise
 
     async def _read_stderr(self, name: str, stderr: asyncio.StreamReader) -> None:

--- a/src/pmcp/transport/http.py
+++ b/src/pmcp/transport/http.py
@@ -19,7 +19,7 @@ import hmac
 import json
 import logging
 import uuid
-from collections.abc import AsyncIterator
+from collections.abc import AsyncIterator, MutableMapping
 from typing import TYPE_CHECKING, Any, Callable
 
 from mcp.server.streamable_http_manager import StreamableHTTPSessionManager
@@ -282,7 +282,7 @@ def create_http_app(
         response_started = False
         original_send = request._send
 
-        async def tracking_send(message: dict[str, Any]) -> None:
+        async def tracking_send(message: MutableMapping[str, Any]) -> None:
             nonlocal response_started
             if message.get("type") == "http.response.start":
                 response_started = True

--- a/src/pmcp/transport/http.py
+++ b/src/pmcp/transport/http.py
@@ -279,10 +279,19 @@ def create_http_app(
 
             request._receive = replay_receive  # type: ignore[method-assign]
 
+        response_started = False
+        original_send = request._send
+
+        async def tracking_send(message: dict[str, Any]) -> None:
+            nonlocal response_started
+            if message.get("type") == "http.response.start":
+                response_started = True
+            await original_send(message)
+
         try:
             await asyncio.wait_for(
                 session_manager.handle_request(
-                    request.scope, request.receive, request._send
+                    request.scope, request.receive, tracking_send
                 ),
                 timeout=request_timeout,
             )
@@ -292,6 +301,8 @@ def create_http_app(
                 request_id,
                 request_timeout,
             )
+            if response_started:
+                return _NullResponse()
             return Response("Gateway Timeout", status_code=504)
         _inc("requests_ok")
         return _NullResponse()

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -1056,6 +1056,127 @@ class TestDoctorAndSecretsIntegration:
         mock_update.assert_awaited_once_with(args)
 
 
+class TestGatewayCliRemoteConfig:
+    """Tests for CLI commands that talk to the running PMCP gateway."""
+
+    def test_get_gateway_url_prefers_new_env_var(
+        self, monkeypatch: pytest.MonkeyPatch
+    ) -> None:
+        """PMCP_GATEWAY_URL should override the legacy SSE alias."""
+        from pmcp.cli import _get_gateway_url
+
+        monkeypatch.setenv("PMCP_STATUS_SSE_URL", "http://legacy.example/mcp")
+        monkeypatch.setenv("PMCP_GATEWAY_URL", "http://gateway.example/mcp")
+
+        assert _get_gateway_url() == "http://gateway.example/mcp"
+
+    @pytest.mark.asyncio
+    async def test_run_update_exits_when_gateway_connect_fails(
+        self, capsys: pytest.CaptureFixture[str]
+    ) -> None:
+        """run_update should stop before tool calls when the gateway is unreachable."""
+        from pmcp.cli import run_update
+
+        args = argparse.Namespace(
+            command="update",
+            server="context7",
+            all=False,
+            json=False,
+            policy=None,
+            log_level="warn",
+        )
+
+        with patch(
+            "pmcp.client.manager.ClientManager.connect_all",
+            new=AsyncMock(return_value=["Failed to connect to pmcp-gateway: boom"]),
+        ) as mock_connect_all:
+            with patch(
+                "pmcp.client.manager.ClientManager.call_tool", new=AsyncMock()
+            ) as mock_call_tool:
+                with patch(
+                    "pmcp.client.manager.ClientManager.disconnect_all",
+                    new=AsyncMock(),
+                ):
+                    with pytest.raises(SystemExit) as exc_info:
+                        await run_update(args)
+
+        assert exc_info.value.code == 1
+        captured = capsys.readouterr()
+        assert "cannot reach PMCP gateway" in captured.err
+        gateway_config = mock_connect_all.await_args.args[0][0]
+        assert gateway_config.config.type == "streamable-http"
+        mock_call_tool.assert_not_awaited()
+
+    @pytest.mark.asyncio
+    async def test_query_running_gateway_status_returns_none_on_connect_failure(
+        self,
+    ) -> None:
+        """Live status query failures should preserve static-config fallback."""
+        from pmcp.cli import _query_running_gateway_status
+
+        class Policy:
+            def get_max_tools_per_server(self) -> int:
+                return 100
+
+        args = argparse.Namespace(server=None, pending=False)
+
+        with patch(
+            "pmcp.client.manager.ClientManager.connect_all",
+            new=AsyncMock(return_value=["Failed to connect to pmcp-gateway: boom"]),
+        ) as mock_connect_all:
+            with patch(
+                "pmcp.client.manager.ClientManager.call_tool", new=AsyncMock()
+            ) as mock_call_tool:
+                with patch(
+                    "pmcp.client.manager.ClientManager.disconnect_all",
+                    new=AsyncMock(),
+                ):
+                    result = await _query_running_gateway_status(args, Policy())
+
+        assert result is None
+        gateway_config = mock_connect_all.await_args.args[0][0]
+        assert gateway_config.config.type == "streamable-http"
+        mock_call_tool.assert_not_awaited()
+
+    @pytest.mark.asyncio
+    async def test_run_auth_connect_exits_when_gateway_connect_fails(
+        self, capsys: pytest.CaptureFixture[str]
+    ) -> None:
+        """auth connect should use streamable HTTP and stop on connect errors."""
+        from pmcp.cli import run_auth_connect
+
+        args = argparse.Namespace(
+            command="auth",
+            auth_command="connect",
+            server_name="browser-use",
+            credential=None,
+            policy=None,
+            log_level="warn",
+            json=False,
+        )
+
+        with patch(
+            "pmcp.client.manager.ClientManager.connect_all",
+            new=AsyncMock(return_value=["Failed to connect to pmcp-gateway: boom"]),
+        ) as mock_connect_all:
+            with patch(
+                "pmcp.client.manager.ClientManager.call_tool", new=AsyncMock()
+            ) as mock_call_tool:
+                with patch(
+                    "pmcp.client.manager.ClientManager.disconnect_all",
+                    new=AsyncMock(),
+                ):
+                    with pytest.raises(SystemExit) as exc_info:
+                        await run_auth_connect(args)
+
+        assert exc_info.value.code == 1
+        captured = capsys.readouterr()
+        assert "cannot reach PMCP gateway" in captured.err
+        gateway_config = mock_connect_all.await_args.args[0][0]
+        assert gateway_config.config.type == "streamable-http"
+        mock_call_tool.assert_not_awaited()
+
+
 class TestRunStatusWithData:
     """Tests for run_status with actual server data."""
 

--- a/tests/test_client_manager.py
+++ b/tests/test_client_manager.py
@@ -359,6 +359,121 @@ class TestRemoteConnectSseHeaders:
 
         await manager.disconnect_all()
 
+    @pytest.mark.asyncio
+    async def test_connect_streamable_http_interpolates_headers_from_env(
+        self, monkeypatch: pytest.MonkeyPatch
+    ) -> None:
+        """Streamable-HTTP remote configs should use the same header interpolation."""
+        manager = ClientManager()
+        monkeypatch.setenv("PMCP_TEST_TOKEN", "test-token")
+
+        config = ResolvedServerConfig(
+            name="remote-http",
+            source="custom",
+            config=RemoteMcpServerConfig(
+                type="streamable-http",
+                url="https://example.com/mcp",
+                headers={
+                    "Authorization": "${PMCP_TEST_TOKEN}",
+                    "X-Static": "literal-value",
+                    "X-Missing": "${PMCP_MISSING}",
+                },
+            ),
+        )
+
+        captured_headers: dict[str, str] = {}
+
+        class EmptyReadStream:
+            def __aiter__(self) -> "EmptyReadStream":
+                return self
+
+            async def __anext__(self) -> None:
+                raise StopAsyncIteration
+
+        @asynccontextmanager
+        async def mock_streamablehttp_client(
+            url: str, headers: dict[str, str] | None = None
+        ):
+            assert url == "https://example.com/mcp"
+            captured_headers.update(headers or {})
+            yield EmptyReadStream(), MagicMock(), MagicMock(return_value=None)
+
+        manager._send_initialize = AsyncMock()
+
+        async def mock_send_request(*args: object, **kwargs: object) -> dict:
+            method = args[1]
+            if method == "tools/list":
+                return {"tools": []}
+            if method == "resources/list":
+                return {"resources": []}
+            if method == "prompts/list":
+                return {"prompts": []}
+            return {}
+
+        manager._send_request = AsyncMock(side_effect=mock_send_request)
+        manager._read_sse = AsyncMock()
+
+        with patch(
+            "pmcp.client.manager.streamablehttp_client", mock_streamablehttp_client
+        ):
+            await manager._connect_streamable_http(config)
+
+        assert captured_headers == {
+            "Authorization": "test-token",
+            "X-Static": "literal-value",
+            "X-Missing": "",
+        }
+
+        await manager.disconnect_all()
+
+
+class TestRemoteConnectTransportDispatch:
+    """Tests for remote transport dispatch."""
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize("config_type", ["http", "streamable-http"])
+    async def test_http_remote_types_use_streamable_http(
+        self, config_type: str
+    ) -> None:
+        """HTTP-style remote configs should use the streamable-HTTP client."""
+        manager = ClientManager()
+        config = ResolvedServerConfig(
+            name=f"remote-{config_type}",
+            source="custom",
+            config=RemoteMcpServerConfig(
+                type=cast(Any, config_type),
+                url="https://example.com/mcp",
+            ),
+        )
+        manager._connect_streamable_http = AsyncMock()  # type: ignore[method-assign]
+        manager._connect_sse = AsyncMock()  # type: ignore[method-assign]
+
+        await manager._connect_server(config)
+
+        manager._connect_streamable_http.assert_awaited_once_with(config)  # type: ignore[attr-defined]
+        manager._connect_sse.assert_not_awaited()  # type: ignore[attr-defined]
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize("config_type", ["sse", "remote"])
+    async def test_legacy_remote_types_use_sse(self, config_type: str) -> None:
+        """SSE and legacy remote configs should keep using the SSE client."""
+        manager = ClientManager()
+        config = ResolvedServerConfig(
+            name=f"remote-{config_type}",
+            source="custom",
+            config=RemoteMcpServerConfig(
+                type=cast(Any, config_type),
+                url="https://example.com/sse",
+            ),
+        )
+        manager._connect_streamable_http = AsyncMock()  # type: ignore[method-assign]
+        manager._connect_sse = AsyncMock()  # type: ignore[method-assign]
+
+        await manager._connect_server(config)
+
+        manager._connect_sse.assert_awaited_once_with(config)  # type: ignore[attr-defined]
+        manager._connect_streamable_http.assert_not_awaited()  # type: ignore[attr-defined]
+
 
 class TestCallTool:
     """Tests for call_tool method."""

--- a/tests/test_transport_http.py
+++ b/tests/test_transport_http.py
@@ -271,6 +271,42 @@ class TestRequestTimeout:
         )
         assert r.status_code == 504
 
+    def test_timeout_after_response_start_does_not_double_send(self) -> None:
+        """A timeout after response start should not trigger a second Starlette response."""
+        mcp_server = MagicMock()
+        mcp_server.list_tools = AsyncMock(return_value=[])
+
+        with patch(
+            "pmcp.transport.http.StreamableHTTPSessionManager",
+            autospec=True,
+        ) as MockManager:
+            instance = MockManager.return_value
+            instance.run.return_value.__aenter__ = AsyncMock(return_value=None)
+            instance.run.return_value.__aexit__ = AsyncMock(return_value=False)
+
+            async def started_then_slow_handler(
+                scope: object, receive: object, send: object
+            ) -> None:
+                await send(
+                    {
+                        "type": "http.response.start",
+                        "status": 202,
+                        "headers": [],
+                    }
+                )
+                await asyncio.sleep(10.0)
+
+            instance.handle_request = started_then_slow_handler
+
+            app = create_http_app(mcp_server, request_timeout=0.05)
+            client = TestClient(app, raise_server_exceptions=False)
+
+        r = client.post(
+            "/mcp",
+            content=b'{"jsonrpc":"2.0","method":"initialize","id":1,"params":{}}',
+        )
+        assert r.status_code == 202
+
     def test_fast_request_does_not_timeout(self) -> None:
         """Request that completes before timeout → not 504."""
         # handler returns immediately (0s delay), timeout is 60s

